### DISCUSSION TITLE
How to restore bootloader on HomeButtons?

### DISCUSSION THREAD

**Discussion by daj102** [nplan/HomeButtons#50] [Q&A]
I was trying to debug why my HB device was not charging the battery ... I noticed that with a laptop USB-C, it was not charging, but when connected to a computer it was connecting fine (so the port isn't broken).  Anyway, short version of my story is that I tried to upload different versions of the firmware to see if it was a problem in the most recent version ... and ended up running `erase_flash` instead of `erase_region` by accident :(

I can put my device into download mode using the buttons, the BOOT LED is on, but despite uploading the firmware correctly - nothing happens.  I'm assuming there is a bootloader which I have now erased/destroyed, and I can't see how to get it back.  Any ideas? @nplan I'm sorry!!!  I destroyed one of your beautiful devices!

**Comment by daj102** [comment]
Well ... I'm still not 100% exactly how I did it, but I saw the note on PlatformIO as the tool used to build the firmware, so I downloaded the code, opened it in VSCode, used PlatformIO to build and then upload to the device ... and I'm back in business :)  I probably need to look into PlatformIO more .... then I can debug my own problems :)

**Comment by nplan** [comment]
Great to hear that you managed to fix it by yourself! You're right - `erase_all` deletes the bootloader and the partition map which are necessary to boot. Uploading with PlatformIO flashes those again.

Just note that downgrading from v2.1.0 won't work for you anymore since you also erased the non volatile storage (NVS) partition that was used to store the factory configuration in earlier versions. In v2.1.0 that was moved to the e-fuses which can't be deleted by esptool commands.

I'll add some info in documentation on how to fix the `erase_all` mistakes. 😉

**Reply by daj102** [reply]
I'm still having my original problem which is that the battery isn't charging ... it obviously charged the battery on the original firmware (2.0.3 I think), but since then I haven't been able to charge at all.  The recent request to see the charging status made me realize that nothing happens when I plug my device in ... I'll do some more digging and see if I can find anything.

**Reply by nplan** [reply]
which hardware revision do you have?

**Reply by daj102** [reply]
v2.3, and I'm running 2.1.0 now ... although I have the full power of PlatformIO now and I did put the 2.1.1 beta on there for a hot second!  I'm learning about how to debug through the code using VSCode, so once I work that out I can dig in.  I noticed that there are flags for battery present, charging, etc ... so I'll try and trace them.

**Reply by nplan** [reply]
What is the voltage of your cell? Charging is only started if it's < 4.0 V. And you have to press a button after connecting USB to trigger charging.

**Reply by daj102** [reply]
Oh interesting ... it's [one of these](https://www.18650batterystore.com/products/panasonic-ncr18650b), so I think the cell voltage should be around 3.6v, but I'll check it with a voltmeter later.  I've noticed that the %age is always around 82-83 percent, even if I leave it plugged in to the charger, so maybe it is just reaching 4v and switching off?  Odd that on the first charge it got to 100% though ...

**Reply by nplan** [reply]
Percentage is calculated directly from voltage. 82% is exactly 4V so I guess this is the issue.

4V is only the threshold to start charging. If the voltage is lower, lets say 3.8V it will trigger charging and charge to full, which is around 4.2V.
